snapcraft/project/_sanity_checks.py

```diff
@@ -45,6 +47,25 @@ def conduct_project_sanity_check(project: snapcraft.project.Project):
     if os.path.isdir(snap_dir_path):
         _check_snap_dir(snap_dir_path)
 
+    _migration_notice(project.info.base)
+
+
+def _migration_notice(base: Optional[str]) -> None:
+    # We only want to encourage developers to use bases if using the snapcraft snap.
+    if base is None and common.is_snap():
+        logger.warning(
+            "This snapcraft project does not use bases which means this project "
+            "does not take advantage of the latest snapcraft features.\n"
+            "Read more about bases at https://docs.snapcraft.io/t/base-snaps/11198"
+        )
+    # Or if they are on the deb (this legacy codebase) and using bases.
+    elif base is not None:
```

Comment:
Does this match the incorrect "base is set AND we're a snap"?

Consider the first `if`: if `base` is set then it is not `None` and so this is false leading us to the next test which is guaranteed to match in this instance because `base` is definitely not `None`. Does this make better sense?

```suggestion
    elif base is not None and not common.is_snap():
```

Reply:
If this is the snap, the 3.x snap, and this situation happens the world will fall apart! There is overly complex re-execution logic (more than necessary) that make it the case that if running from the snap, you will most likely not fall into this situation.

I guess this suggestion could tie it to the comment above.